qiskit/primitives/base_sampler.py

```diff
@@ -383,6 +384,23 @@ def run(
                     f"The number of values ({len(parameter_value)}) does not match "
                     f"the number of parameters ({circuit.num_parameters}) for the {i}-th circuit."
                 )
+
+        for i, circuit in enumerate(circuits):
+            if circuit.num_clbits == 0:
+                raise QiskitError(
+                    f"The {i}-th circuit does not have any classical bit. "
+                    "Sampler requires classical bits, plus measurements "
+                    "on the desired qubits."
+                )
+
+            mapping = final_measurement_mapping(circuit)
+            if set(range(circuit.num_clbits)) != set(mapping.values()):
```

Comment:
I am not clear why the number of final measurements needs to equal the number of bits.  For example, what if I use mid-circuit measurements to write to my register?

Reply:
The reference implementation originally did not support mid-circuits measurements because it needs a qubit mapping and uses `final_measurement_mapping`. So, I moved the check to the base interface in this PR.

I try to think of how to support mid-circuit measurement for `Sampler`. What is the best way to get the qubit mapping of a circuit with mid-circuit measurement?

Reply:
So things become a bit more complicated in that case.  I am working on a related feature in M3 for mid-circuit measurement mitigation, but there is no ideal automated solution when dynamic circuits get involved.  I think the M3 tool can be used when I am done with it (hopefully next week).  Until then I would just check if any measurements at all and/or the presence of clbits.

Reply:
Thanks. This PR has a check the presence of clbits already. I will make another PR to update the check as follows to see whether there is any measurement or not.

```python
    if not final_measurement_mapping(circuit):
       raise QIskitError("no measurement at all")
```

Reply:
But that is still not correct.  I can have mid-circuit measurements.  Perhaps just leave a classical register check for now

Reply:
Thank you again. I didn't know it. Mid-circuit measurement is more tricky than I thought. I agree on your idea to check just classical registers.

```python
from qiskit import QuantumCircuit
from mthree.utils import final_measurement_mapping

qc = QuantumCircuit(1, 1)
qc.measure(0, 0)
qc.h(0)
print(qc)
print(final_measurement_mapping(qc))
```

output
```
     ┌─┐┌───┐
  q: ┤M├┤ H ├
     └╥┘└───┘
c: 1/═╩══════
      0
{}
```